File: mesh-delivery/android/THEOplayerMeshDeliveryJava/app/src/main/java/io/streamroot/lumen/delivery/client/samples/theoplayer/PlayerActivity.java
package io.streamroot.lumen.delivery.client.samples.theoplayer;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import com.theoplayer.android.api.THEOplayerConfig;
import com.theoplayer.android.api.THEOplayerView;
import com.theoplayer.android.api.player.Player;
import com.theoplayer.android.api.source.SourceDescription;
import com.theoplayer.android.api.source.TypedSource;
import com.theoplayer.android.api.ui.UIConfiguration;
import io.streamroot.lumen.delivery.client.core.LumenDeliveryClient;
import io.streamroot.lumen.delivery.client.samples.theoplayer.databinding.ActivityPlayerBinding;
import io.streamroot.lumen.delivery.client.utils.LumenStatsView;

public class PlayerActivity extends AppCompatActivity {
  private static final String ARG_STREAM_CONFIG = "streamConfig";

  public static Intent makeIntent(Context ctx, @NonNull StreamConfig config) {
    return new Intent(ctx, PlayerActivity.class).putExtra(ARG_STREAM_CONFIG, config);
  }

  private static @NonNull StreamConfig extractConfig(Intent i) {
    return i.getParcelableExtra(ARG_STREAM_CONFIG);
  }

  private @NonNull StreamConfig streamConfig;

  private @NonNull ActivityPlayerBinding bindings;
  private @Nullable THEOplayerView player = null;
  private @Nullable LumenDeliveryClient lumenDeliveryClient = null;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    bindings = ActivityPlayerBinding.inflate(getLayoutInflater());
    setContentView(bindings.getRoot());

    streamConfig = extractConfig(getIntent());
  }

  @Override
  protected void onStart() {
    super.onStart();
    if (Build.VERSION.SDK_INT > 23) {
      initializePlayer();
    }
  }

  @Override
  protected void onResume() {
    super.onResume();
    if (Build.VERSION.SDK_INT <= 23 || player == null) {
      initializePlayer();
    }
    if (player != null) {
      player.onResume();
    }
  }

  @Override
  protected void onPause() {
    if (Build.VERSION.SDK_INT <= 23) {
      releasePlayer();
    }
    if (player != null) {
      player.onPause();
    }
    super.onPause();
  }

  @Override
  protected void onStop() {
    if (Build.VERSION.SDK_INT > 23) {
      releasePlayer();
    }
    super.onStop();
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    if (player != null) {
      player.onDestroy();
    }
  }

  private void initializePlayer() {
    if (player == null) {
      bindings.playerContainer.removeAllViews();
      THEOplayerConfig playerConfig =
          new THEOplayerConfig.Builder()
              .license(Application.THEO_LICENSE(this))
              .build();
      player = new THEOplayerView(this, playerConfig);

      bindings.playerContainer.addView(
          player, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

      if (streamConfig.isStreamrootEnabled) {
        // Below, we initialize and start Lumen delivery client
        lumenDeliveryClient = createDeliveryClient(player.getPlayer());
        lumenDeliveryClient.start();
        showStatsView(lumenDeliveryClient);
      }

      // We fetch the final URL from the delivery client
      final String deliveryManifest =
          lumenDeliveryClient != null ? lumenDeliveryClient.localUrl() : streamConfig.contentUrl;

      TypedSource typedSource = TypedSource.Builder.typedSource().src(deliveryManifest).build();

      SourceDescription sourceDescription =
          SourceDescription.Builder.sourceDescription(typedSource).build();

      player.getPlayer().setSource(sourceDescription);
      if (streamConfig.autostart) {
        player.getPlayer().play();
      }

      // Add controls for unified only
      PlayerControls.setupUIForUnified(
          getLayoutInflater(), player.getPlayer(), bindings.playerContainer);
    }
  }

  private LumenDeliveryClient createDeliveryClient(Player player) {
    final PlayerInteractor playerInteractor = new PlayerInteractor(player);

    return LumenDeliveryClient.meshBuilder(this)
        .playerInteractor(playerInteractor)
        .options(
            o -> {
              o.logLevel(streamConfig.logLevel);
              o.deliveryClientKey(streamConfig.streamrootKey);
              if (streamConfig.property != null) o.meshProperty(streamConfig.property);
              if (streamConfig.contentId != null) o.contentId(streamConfig.contentId);
              if (streamConfig.latency != null) o.latency(Short.parseShort(streamConfig.latency));
              return null;
            })
        .build(streamConfig.contentUrl);
  }

  private void showStatsView(LumenDeliveryClient dc) {
    ViewGroup svc = bindings.statsviewContainer;
    svc.removeAllViews();
    final LumenStatsView statsView = new LumenStatsView(this);
    dc.addStateStatsListener(statsView);
    statsView.showStats();
    svc.addView(
        statsView, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
  }

  private void releasePlayer() {
    if (player == null) return;

    stopDeliveryClient();

    player.getPlayer().stop();
    player = null;
  }

  private void stopDeliveryClient() {
    if (lumenDeliveryClient == null) return;
    lumenDeliveryClient.stop();
    lumenDeliveryClient = null;
  }
}

File: mesh-delivery/android/ExoPlayerMeshDeliveryPlugin/app/src/main/java/io/streamroot/lumen/delivery/client/samples/mesh_delivery/exoplayer/PlayerActivity.kt
package io.streamroot.lumen.delivery.client.samples.mesh_delivery.exoplayer

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import com.google.android.exoplayer2.MediaItem
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.util.Util
import io.streamroot.lumen.delivery.client.samples.mesh_delivery.exoplayer.databinding.ActivityPlayerBinding
import io.streamroot.lumen.delivery.client.core.LumenLogLevel
import io.streamroot.lumen.delivery.client.plugin.LumenDeliveryClientPlugin
import io.streamroot.lumen.delivery.client.utils.LumenStatsView

class PlayerActivity : AppCompatActivity() {
    data class PlayerActivityArgs(
        val deliveryClientKey: String?,
        val streamUrl: String?,
        val meshProperty: String?
    )

    companion object {
        private const val ARG_DC_KEY = "deliveryClientKey"
        private const val ARG_STREAM_URL = "streamUrl"
        private const val ARG_MESH_PROP = "meshProperty"

        fun makeIntent(ctx: Context, args: PlayerActivityArgs): Intent {
            return Intent(ctx, PlayerActivity::class.java).apply {
                putExtra(ARG_STREAM_URL, args.streamUrl)
                putExtra(ARG_MESH_PROP, args.meshProperty)
                putExtra(ARG_DC_KEY, args.deliveryClientKey)
            }
        }

        fun extractArgs(i: Intent): PlayerActivityArgs {
            return PlayerActivityArgs(
                i.getStringExtra(ARG_DC_KEY),
                i.getStringExtra(ARG_STREAM_URL),
                i.getStringExtra(ARG_MESH_PROP)
            )
        }
    }

    private var deliveryClientKey: String? = null
    private lateinit var streamUrl: String
    private var meshProperty: String? = null

    private lateinit var bindings: ActivityPlayerBinding
    private var plugin: LumenDeliveryClientPlugin? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bindings = ActivityPlayerBinding.inflate(layoutInflater)
        setContentView(bindings.root)

        val args = extractArgs(intent)
        streamUrl = args.streamUrl!!
        deliveryClientKey = args.deliveryClientKey?.takeUnless { it.isBlank() }
        meshProperty = args.meshProperty?.takeUnless { it.isBlank() }
    }

    override fun onStart() {
        super.onStart()
        if (Util.SDK_INT > 23) {
            initializePlayer()
        }
    }

    override fun onResume() {
        super.onResume()
        if (Util.SDK_INT <= 23 || plugin == null) {
            initializePlayer()
        }
    }

    override fun onPause() {
        if (Util.SDK_INT <= 23) {
            releasePlayer()
        }
        super.onPause()
    }

    override fun onStop() {
        if (Util.SDK_INT > 23) {
            releasePlayer()
        }
        super.onStop()
    }

    private fun initializePlayer() {
        if (plugin == null) {
            LumenDeliveryClientPlugin.Builder(this, streamUrl).meshOptions {
                deliveryClientKey?.let { deliveryClientKey(it) }
                meshProperty?.let { meshProperty(it) }
                logLevel(LumenLogLevel.INFO)
            }.start().apply {
                plugin = this

                bindings.playerView.player = exoPlayer
                showStatsView(this)

                exoPlayer.playWhenReady = true
                exoPlayer.addMediaItem(MediaItem.fromUri(Uri.parse(finalUri)))
                exoPlayer.repeatMode = Player.REPEAT_MODE_ALL
                exoPlayer.prepare()
            }
        }
    }

    private fun showStatsView(dc: LumenDeliveryClientPlugin) {
        bindings.statsviewContainer.apply {
            removeAllViews()
            val statsView = LumenStatsView(this@PlayerActivity)
            dc.addStateStatsListener(statsView)
            statsView.showStats()
            addView(
                statsView,
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
        }
    }

    private fun releasePlayer() {
        if (plugin == null) return

        stopDeliveryClient()
    }

    private fun stopDeliveryClient() {
        plugin?.stop()
        plugin?.exoPlayer?.release()
        plugin = null
    }
}